### PR TITLE
Hdrp/feature/material blend state

### PR DESCRIPTION
**Pull request by alelievr**
### Purpose of this PR
Add the control over `ZTest`, `ZWrite` and `CullMode` on transparent materials (include shader graph ones)

image
image

Note: The `CullMode` is only visible in the inspector when bouble sided is off

These option are also displayed on HDRP master nodes in shader graph (PBR master node not included) and allow you to set the default parameters of these blend states when a material is created from a shader (right click on the shader and then create a material).
image

---
### Testing status
**Katana Tests**: https://katana.bf.unity3d.com/projects/com.unity.render-pipelines/builders?ScriptableRenderLoop_branch=HDRP%2Ffeature%2Fmaterial_blend_state&automation-tools_branch=master&unity_branch=trunk

**Manual Tests**: What did you do?
- [ ] Opened test project + Run graphic tests locally
- [ ] Built a player
- [ ] Checked new UI names with UX convention
- [ ] Tested UI multi-edition + Undo/Redo
- [ ] C# and shader warnings (supress shader cache to see them)
- [ ] Checked new resources path for the reloader (in devloper mode, you have a button at end of resources that check the pathes)

Graphic test added: 
8105_BlendStates
image

---
### Overall Product Risks
**Technical Risk**: Low

**Halo Effect**: Low

---
### Comments to reviewers

Note: this PR include change from this one https://github.com/Unity-Technologies/ScriptableRenderPipeline/pull/3333

### CODE REVIEW

File: TestProjects/HDRP_Tests/Assets/Temporary.meta [hunk 1]
```diff
@@ -0,0 +1,8 @@
+fileFormatVersion: 2
+guid: aecca44940d719744ba8063d12df690b
```
Comment (sebastienlagarde) [inline TestProjects/HDRP_Tests/Assets/Temporary.meta]: expected temporary.meta ?

File: com.unity.render-pipelines.high-definition/CHANGELOG.md [hunk 2]
```diff
@@ -10,6 +10,9 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
 - `Fixed`, `Viewer`, and `Automatic` modes to compute the FOV used when rendering a `PlanarReflectionProbe`
 - A checkbox to toggle the chrome gizmo of `ReflectionProbe`and `PlanarReflectionProbe`
 - Added a Light layer in shadows that allow for objects to cast shadows without being affected by light (and vice versa).
+- ShaderGraph blend states are now accessible from the material UI (Surface Type, Sorting Priority, Blending Mode, etc.). Materials using a ShaderGraph may be broken after this change so we provide an upgrade script for the materials in the scene: https://gist.github.com/alelievr/8003b0892558deee911cd4d6189d68e4
```
Comment (sebastienlagarde) [inline com.unity.render-pipelines.high-definition/CHANGELOG.md]: no: provide the command to execute the script :)

File: com.unity.render-pipelines.high-definition/CHANGELOG.md [hunk 3]
```diff
@@ -10,6 +10,9 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
...
+- Added the control over ZTest, ZWrite and CullMode for transparent materials
+- Unlit HD Unlit master node shaders now cast shadows, as well as the Unlit Shader
```
Comment (sebastienlagarde) [inline com.unity.render-pipelines.high-definition/CHANGELOG.md]: Where this come from?
" Unlit HD Unlit master node shaders now cast shadows, as well as the Unlit Shader"
if we have it it should be an option, don't want to the UI to cast shadow in the world :). But currently I am not sure if we have such an option or not.

File: com.unity.render-pipelines.high-definition/Editor/Material/Fabric/ShaderGraph/FabricPass.template [hunk 4]
```diff
@@ -29,17 +29,17 @@ Pass
 
     #pragma multi_compile _ LOD_FADE_CROSSFADE
 
+    #pragma shader_feature _SURFACE_TYPE_TRANSPARENT
+    #pragma shader_feature_local _DOUBLESIDED_ON
+    #pragma shader_feature_local _ _BLENDMODE_ALPHA _BLENDMODE_ADD _BLENDMODE_PRE_MULTIPLY
+    #pragma shader_feature_local _BLENDMODE_PRESERVE_SPECULAR_LIGHTING
```
Comment (sebastienlagarde) [inline com.unity.render-pipelines.high-definition/Editor/Material/Fabric/ShaderGraph/FabricPass.template]: Cautoin: to remove _BLENDMODE_PRESERVE_SPECULAR_LIGHTING

File: com.unity.render-pipelines.high-definition/Editor/Material/Hair/ShaderGraph/HairPass.template [hunk 5]
```diff
@@ -29,15 +29,15 @@ Pass
 
     #pragma multi_compile _ LOD_FADE_CROSSFADE
 
+    #pragma shader_feature_local _ _BLENDMODE_ALPHA _BLENDMODE_ADD _BLENDMODE_PRE_MULTIPLY
+    #pragma shader_feature_local _BLENDMODE_PRESERVE_SPECULAR_LIGHTING
```
Comment (sebastienlagarde) [inline com.unity.render-pipelines.high-definition/Editor/Material/Hair/ShaderGraph/HairPass.template]: do a pass on all _BLENDMODE_PRESERVE_SPECULAR_LIGHTING

File: com.unity.render-pipelines.high-definition/Editor/Material/Lit/LitShaderPreprocessor.cs [hunk 6]
```diff
@@ -25,6 +26,15 @@ public override bool ShadersStripper(HDRenderPipelineAsset hdrpAsset, Shader sha
             // Using Contains to include the Tessellation variants
             bool isBuiltInLit = shader.name.Contains("HDRP/Lit") || shader.name.Contains("HDRP/LayeredLit") || shader.name.Contains("HDRP/TerrainLit");
 
+            if (shader.IsShaderGraph())
+            {
+                string shaderPath = AssetDatabase.GetAssetPath(shader);
+                isBuiltInLit |= GraphUtil.GetOutputNodeType(shaderPath) == typeof(HDLitMasterNode);
+                isBuiltInLit |= GraphUtil.GetOutputNodeType(shaderPath) == typeof(HairMasterNode);
```
Comment (sebastienlagarde) [inline com.unity.render-pipelines.high-definition/Editor/Material/Lit/LitShaderPreprocessor.cs]: removed 	                isBuiltInLit |= GraphUtil.GetOutputNodeType(shaderPath) == typeof(HairMasterNode);
                isBuiltInLit |= GraphUtil.GetOutputNodeType(shaderPath) == typeof(FabricMasterNode);
                isBuiltInLit |= GraphUtil.GetOutputNodeType(shaderPath) == typeof(StackLitMasterNode);

only HDLitMasterNode is qrequried here. We want to know which one use deferred

File: com.unity.shadergraph/Editor/Drawing/PreviewManager.cs [hunk 7]
```diff
@@ -725,4 +725,4 @@ public void NotifyPreviewChanged()
                 onPreviewChanged();
         }
     }
-}
+}
```
Comment (sebastienlagarde) [inline com.unity.shadergraph/Editor/Drawing/PreviewManager.cs]: ?

File: com.unity.shadergraph/Editor/Importers/ShaderGraphImporter.cs [hunk 8]
```diff
@@ -146,4 +146,4 @@ internal static string GetShaderText(string path, out List<PropertyCollector.Tex
             return GetShaderText(path, out configuredTextures, null, out _);
         }
     }
```
Comment (sebastienlagarde) [inline com.unity.shadergraph/Editor/Importers/ShaderGraphImporter.cs]: ?

File: com.unity.shadergraph/Editor/Data/Nodes/MasterNode.cs [hunk 9]
```diff
@@ -6,6 +6,7 @@
 using UnityEngine;
 using UnityEngine.Rendering;
 using UnityEngine.UIElements;
+using UnityEditor.ShaderGraph.Drawing.Inspector;
```
Comment (sebastienlagarde) [inline com.unity.shadergraph/Editor/Data/Nodes/MasterNode.cs]: is this really required? there is no other change here

File: com.unity.shadergraph/Editor/Data/Nodes/IMasterNode.cs [hunk 10]
```diff
@@ -1,6 +1,7 @@
 using System.Collections.Generic;
 using UnityEngine;
 using UnityEngine.Rendering;
+using UnityEditor.ShaderGraph.Drawing.Inspector;
```
Comment (sebastienlagarde) [inline com.unity.shadergraph/Editor/Data/Nodes/IMasterNode.cs]: is this really required? there is no other change here

File: com.unity.render-pipelines.high-definition/Runtime/Material/MaterialExtension.cs [hunk 11]
```diff
@@ -0,0 +1,154 @@
...
+        None
+    }
+
+    public enum TessellationMode
+    {
+        None,
+        Phong
+    }
+
+    public enum MaterialId
```
Comment (sebastienlagarde) [inline com.unity.render-pipelines.high-definition/Runtime/Material/MaterialExtension.cs]: Does it still make sense to have it here? this isn't a material extension anymore right ?
Maybe move it again to a lit MAterial ?

File: com.unity.render-pipelines.high-definition/Runtime/Material/MaterialExtension.cs [hunk 12]
```diff
@@ -0,0 +1,154 @@
...
+    }
+
+    public enum VertexColorMode
+    {
+        None,
+        Multiply,
+        Add
+    }
+
+    internal enum    AxfBrdfType
```
Comment (sebastienlagarde) [inline com.unity.render-pipelines.high-definition/Runtime/Material/MaterialExtension.cs]: Why do we need AxF definition here ?

File: com.unity.render-pipelines.high-definition/Runtime/Material/Lit/LitTessellation.shader [hunk 13]
```diff
@@ -155,12 +155,14 @@ Shader "HDRP/LitTessellation"
         [HideInInspector] _DstBlend("__dst", Float) = 0.0
         [HideInInspector] _AlphaSrcBlend("__alphaSrc", Float) = 1.0
         [HideInInspector] _AlphaDstBlend("__alphaDst", Float) = 0.0
-        [HideInInspector] _ZWrite("__zw", Float) = 1.0
+        [HideInInspector][ToggleUI] _ZWrite("__zw", Float) = 1.0
         [HideInInspector] _CullMode("__cullmode", Float) = 2.0
         [HideInInspector] _CullModeForward("__cullmodeForward", Float) = 2.0 // This mode is dedicated to Forward to correctly handle backface then front face rendering thin transparent
+        [Enum(Front, 1, Back, 2)] _TransparentCullMode("_TransparentCullMode", Int) = 2 // Back culling by default
```
Comment (sebastienlagarde) [inline com.unity.render-pipelines.high-definition/Runtime/Material/Lit/LitTessellation.shader]: I see below that you used an enum, can we do the same here? with TransparentCullMode ?

File: com.unity.render-pipelines.high-definition/Runtime/Lighting/LightLoop/LightLoop.cs [hunk 14]
```diff
@@ -2844,7 +2844,6 @@ public bool WillRenderScreenSpaceShadows()
 #endif
         }
```
Comment (sebastienlagarde) [inline com.unity.render-pipelines.high-definition/Runtime/Lighting/LightLoop/LightLoop.cs]: not needed (will avoid you conflict :) )

File: com.unity.render-pipelines.high-definition/Editor/RenderPipeline/HDEditorUtils.cs [hunk 15]
```diff
@@ -20,20 +21,60 @@ public class HDEditorUtils
...
+            { "HDRP/Unlit", UnlitGUI.SetupUnlitMaterialKeywordsAndPass },
             { "HDRP/Decal", DecalUI.SetupMaterialKeywordsAndPass },
-            { "HDRP/TerrainLit", TerrainLitGUI.SetupMaterialKeywordsAndPass }
+            { "HDRP/TerrainLit", TerrainLitGUI.SetupMaterialKeywordsAndPass },
+            { "HDRP/AxF", AxFGUI.SetupMaterialKeywordsAndPass }
+        };
+
+        static Dictionary<Type, MaterialResetter> k_ShaderGraphMaterialResetters = new Dictionary<Type, MaterialResetter>
+        {
+            { typeof(HDUnlitMasterNode), UnlitGUI.SetupUnlitMaterialKeywordsAndPass },
```
Comment (sebastienlagarde) [inline com.unity.render-pipelines.high-definition/Editor/RenderPipeline/HDEditorUtils.cs]: There is also the PBR and Unlit shader graph (the one that are cross pipeline)

File: com.unity.render-pipelines.high-definition/Editor/Material/Unlit/ShaderGraph/UnlitPass.template [hunk 16]
```diff
@@ -146,4 +146,4 @@ $splice(Includes)
     //-------------------------------------------------------------------------------------
```
Comment (sebastienlagarde) [inline com.unity.render-pipelines.high-definition/Editor/Material/Unlit/ShaderGraph/UnlitPass.template]: end of line?

File: com.unity.render-pipelines.high-definition/Editor/Material/Unlit/ShaderGraph/HDUnlitPass.template [hunk 17]
```diff
@@ -24,17 +24,22 @@ Pass
...
-    $BlendMode.Premultiply:              #define _BLENDMODE_PRE_MULTIPLY 1
-    $AlphaFog:                           #define _ENABLE_FOG_ON_TRANSPARENT 1
+    
+    #pragma shader_feature_local _ALPHATEST_ON
+    // #pragma shader_feature_local _DOUBLESIDED_ON - We have no lighting, so no need to have this combination for shader, the option will just disable backface culling
+
+    #pragma shader_feature_local _EMISSIVE_COLOR_MAP
+
+    // Keyword for transparent
+    #pragma shader_feature _SURFACE_TYPE_TRANSPARENT
```
Comment (sebastienlagarde) [inline com.unity.render-pipelines.high-definition/Editor/Material/Unlit/ShaderGraph/HDUnlitPass.template]: Please be sure that all    _SURFACE_TYPE_TRANSPARENT use a shader_feature and not a loca_shader _feature (not sure if it is the case everywhere, so I prefer to warn it here);
 #pragma shader_feature _SURFACE_TYPE_TRANSPARENT

File: com.unity.render-pipelines.high-definition/Editor/Material/Unlit/ShaderGraph/HDUnlitPass.template [hunk 18]
```diff
@@ -24,17 +24,22 @@ Pass
...
-    $SurfaceType.Transparent:            #define _SURFACE_TYPE_TRANSPARENT 1
-    $BlendMode.Alpha:                    #define _BLENDMODE_ALPHA 1
-    $BlendMode.Add:                      #define _BLENDMODE_ADD 1
-    $BlendMode.Premultiply:              #define _BLENDMODE_PRE_MULTIPLY 1
-    $AlphaFog:                           #define _ENABLE_FOG_ON_TRANSPARENT 1
+    
+    #pragma shader_feature_local _ALPHATEST_ON
+    // #pragma shader_feature_local _DOUBLESIDED_ON - We have no lighting, so no need to have this combination for shader, the option will just disable backface culling
+
+    #pragma shader_feature_local _EMISSIVE_COLOR_MAP
```
Comment (sebastienlagarde) [inline com.unity.render-pipelines.high-definition/Editor/Material/Unlit/ShaderGraph/HDUnlitPass.template]: It is not expected to have this shader feature for emissive here
_EMISSIVE_COLOR_MAP

File: com.unity.render-pipelines.high-definition/Editor/Material/Unlit/ShaderGraph/HDUnlitPass.template [hunk 19]
```diff
@@ -24,17 +24,22 @@ Pass
...
+
+    // Keyword for transparent
+    #pragma shader_feature _SURFACE_TYPE_TRANSPARENT
+    #pragma shader_feature_local _ _BLENDMODE_ALPHA _BLENDMODE_ADD _BLENDMODE_PRE_MULTIPLY
+    #pragma shader_feature_local _ENABLE_FOG_ON_TRANSPARENT
```
Comment (sebastienlagarde) [inline com.unity.render-pipelines.high-definition/Editor/Material/Unlit/ShaderGraph/HDUnlitPass.template]: hmm... I don't think we have expose fog on transparent, do we have?
_ENABLE_FOG_ON_TRANSPARENT
This is not requried

**Comment by sebastienlagarde** [comment]
@RemyUnity Be sure to have your team testing multiselection and undo/redo. All material need to be manipulated in UI with the various mode (refractoin, distortion, material Id etc...) as well as all the shader graph...

File: TestProjects/HDRP_Tests/Assets/Temporary.meta (hunk 1 above)
Comment (alelievr) [reply TestProjects/HDRP_Tests/Assets/Temporary.meta]: No, I'll delete it

File: com.unity.render-pipelines.high-definition/CHANGELOG.md (hunk 2 above)
Comment (alelievr) [reply com.unity.render-pipelines.high-definition/CHANGELOG.md]: So I put the upgrade script in the UpgradeMenuIterm.cs file ? with a `MenuItem`

File: com.unity.render-pipelines.high-definition/Runtime/Material/MaterialExtension.cs (hunk 12 above)
Comment (alelievr) [reply com.unity.render-pipelines.high-definition/Runtime/Material/MaterialExtension.cs]: will move to AxfGUI.cs

File: com.unity.render-pipelines.high-definition/Editor/RenderPipeline/HDEditorUtils.cs (hunk 15 above)
Comment (alelievr) [reply com.unity.render-pipelines.high-definition/Editor/RenderPipeline/HDEditorUtils.cs]: But they don't need their keywords to be reseted right ?

File: com.unity.render-pipelines.high-definition/Editor/Material/Unlit/ShaderGraph/HDUnlitPass.template (hunk 17 above)
Comment (alelievr) [reply com.unity.render-pipelines.high-definition/Editor/Material/Unlit/ShaderGraph/HDUnlitPass.template]: I just checked, all the `_SURFACE_TYPE_TRANSPARENT` are using `shader_feature`

File: com.unity.render-pipelines.high-definition/Editor/Material/Unlit/ShaderGraph/HDUnlitPass.template (hunk 19 above)
Comment (alelievr) [reply com.unity.render-pipelines.high-definition/Editor/Material/Unlit/ShaderGraph/HDUnlitPass.template]: We haven't, as well as the AlphaTest, will remove them

File: com.unity.render-pipelines.high-definition/Editor/Material/Unlit/ShaderGraph/HDUnlitPass.template (hunk 18 above)
Comment (alelievr) [reply com.unity.render-pipelines.high-definition/Editor/Material/Unlit/ShaderGraph/HDUnlitPass.template]: will remove

File: com.unity.render-pipelines.high-definition/Runtime/Material/Lit/LitTessellation.shader (hunk 13 above)
Comment (alelievr) [reply com.unity.render-pipelines.high-definition/Runtime/Material/Lit/LitTessellation.shader]: done

File: com.unity.render-pipelines.high-definition/Editor/Material/Fabric/ShaderGraph/FabricPass.template (hunk 4 above)
Comment (alelievr) [reply com.unity.render-pipelines.high-definition/Editor/Material/Fabric/ShaderGraph/FabricPass.template]: done

File: com.unity.render-pipelines.high-definition/CHANGELOG.md [hunk 20]
```diff
@@ -10,6 +10,9 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
 - `Fixed`, `Viewer`, and `Automatic` modes to compute the FOV used when rendering a `PlanarReflectionProbe`
 - A checkbox to toggle the chrome gizmo of `ReflectionProbe`and `PlanarReflectionProbe`
 - Added a Light layer in shadows that allow for objects to cast shadows without being affected by light (and vice versa).
+- ShaderGraph blend states are now accessible from the material UI (Surface Type, Sorting Priority, Blending Mode, etc.). Materials using a ShaderGraph may be broken after this change, to fix them you can use the `Edit/Render Pipeline/Reset all ShaderGraph Scene Materials BlendStates` option that will synch the blendstate of your shader graph master node with the material properties.
```
Comment (JordanL8) [inline com.unity.render-pipelines.high-definition/CHANGELOG.md]: You can now access ShaderGraph blend states from the Material UI (for example, **Surface Type**, **Sorting Priority**, and **Blending Mode**). This change may break Materials that use a ShaderGraph, to fix them, select **Edit > Render Pipeline > Reset all ShaderGraph Scene Materials BlendStates**. This syncs the blendstates of you ShaderGraph master nodes with the Material properties.

File: com.unity.render-pipelines.high-definition/CHANGELOG.md [hunk 21]
```diff
@@ -10,6 +10,9 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
...
+- Added the control over ZTest, ZWrite and CullMode for transparent materials
```
Comment (JordanL8) [inline com.unity.render-pipelines.high-definition/CHANGELOG.md]: You can now control ZTest, ZWrite, and CullMode for transparent Materials.

File: com.unity.render-pipelines.high-definition/CHANGELOG.md [hunk 22]
```diff
@@ -10,6 +10,9 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
...
+- Unlit HD Unlit master node shaders now cast shadows, as well as the Unlit Shader
```
Comment (JordanL8) [inline com.unity.render-pipelines.high-definition/CHANGELOG.md]: Materials that use Unlit Shaders or Unlit Master Node Shaders now cast shadows.

File: com.unity.render-pipelines.high-definition/CHANGELOG.md [hunk 23]
```diff
@@ -30,6 +33,7 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
 - Replaced HDCamera.computePassCount by viewCount
 - Removed xrInstancing flag in RTHandles (replaced by TextureXR.slices and TextureXR.dimensions)
 - Refactor the HDRenderPipeline and lightloop code to preprare for high level rendergraph
+- Removed the Back Then Front Rendering option in the fabric master node settings that didn't do anything
```
Comment (JordanL8) [inline com.unity.render-pipelines.high-definition/CHANGELOG.md]: Removed the **Back Then Front Rendering** option in the fabric Master Node settings. Enabling this option previously did nothing.

File: TestProjects/HDRP_Tests/Assets/GraphicTests/Common/Scripts/Editor/ResetShaderGraphMaterialProperties.cs [hunk 24]
```diff
@@ -0,0 +1,59 @@
...
+        "_DoubleSidedEnable", "_DoubleSidedNormalMode",
+        "_TransparentBackfaceEnable", "_ReceivesSSR", "_RequireSplitLighting"
+    };
+
+    static readonly string[] vectorPropertiesToReset = {
+        "_DoubleSidedConstants",
+    };
+
+    [MenuItem("Edit/Render Pipeline/Reset All ShaderGraph Material Properties %g")]
+    static void ResetShaderGraphMaterialProperties()
```
Comment (JulienIgnace-Unity) [inline TestProjects/HDRP_Tests/Assets/GraphicTests/Common/Scripts/Editor/ResetShaderGraphMaterialProperties.cs]: Maybe a small comment to explain why this is done this way?

Comment (alelievr) [reply TestProjects/HDRP_Tests/Assets/GraphicTests/Common/Scripts/Editor/ResetShaderGraphMaterialProperties.cs]: done

**Pull request merged**